src/FileSig.hs

```diff
@@ -7,126 +7,143 @@
-type Signature = (String, Int)
+import Data.List as L (sort)
+import Data.Ord (Down(..))
+import Data.Text as T (Text, pack, unlines, unpack)
+import System.IO (IOMode(..), hClose, withBinaryFile)
+
+
+import Magic
+
+
+-- | Magic bytes/file signature, in the form of the offset (`Int`), followed
+-- by the bytes (`ByteString`).
+type Signature = (Int, ByteString)
 
 -- | The extension of a file format, which acts as a key in the magic database.
 type ExtensionKey = Text
 
+-- | The MIME type/media type
+type MIME = Text
+
 -- | File format located by a file extension (as key) in the magic number database (JSON).
 data FileFormat = FileFormat {
   magicBytes :: [Signature],
   -- ^ The magic numbers/file signatures which can be used to identify this file format.
-  mime :: Text
+  mime :: MIME
   -- ^ the MIME type/media type of this file format.
 } deriving (Show)
 
 -- | The abstraction of the JSON magic database containing all the file signatures.
 -- The key is the file extension.
 type MagicMap = HM.HashMap ExtensionKey FileFormat
 
--- | JSON parser for a file format entries in the magic database.
-instance FromJSON FileFormat where
-  parseJSON = withObject "FileFormat" $ \o ->
-    FileFormat <$> fmap parseSignatures (o .: "signs")
-            <*> o .: "mime"
+-- | This is the entire MagicMap to be used everywhere, no IO required.
+--
+-- The 'decodeBase16Lenient' interprets the hexidecimal into raw bytestring,
+-- So it is already in optimal form at runtime.
+magicMap :: MagicMap
+magicMap =
+    HM.map go rawMagicMap
+  where
+    go :: ([Signature], Text) -> FileFormat
+    go (sigs, mime) = FileFormat (fmap decodeBase16Lenient <$> sigs) mime
+
+-- | All extensions in the 'magicMap'
+allExtensions :: [ExtensionKey]
+allExtensions = L.sort $ HM.keys magicMap
+
+-- This is going to be '(36865, "4344303031")' from the 'iso' signature
+-- but if the magicMap is ever updated, this will consistently give the
+-- biggest starting chunk we might need to guess extentions, so as to not
+-- load in potential GBs of data into memory.
+longestSignature :: Int
+longestSignature = longestChunk allSigs
+  where
+    allSigs = concatMap magicBytes $ HM.elems magicMap
+
+sigLength :: Signature -> Int
+sigLength (i, bs) = i + BS.length bs
 
--- | Parse the JSON "signs" entry from the magic database.
-parseSignatures :: Value -> [Signature]
-parseSignatures someTexts@(Array _) = L.map parseSig (listOfText someTexts)
+getLongestSignature :: ExtensionKey -> Maybe Int
+getLongestSignature ext =
+    longestChunk . magicBytes <$> mFileExt
   where
-    listOfText (Array a) =
-      let blah (String x) = x
-          aList = V.toList a
-      in  L.map blah aList
-
-    parseSig :: Text -> Signature
-    parseSig sign =
-      let [offset, hex] = T.splitOn "," sign
-      in  (T.unpack hex, read (T.unpack offset) :: Int)
-parseSignatures _ = mzero
-
-getJSON :: B.ByteString
-getJSON = do
-  B.fromStrict $(embedFile "data/magic.json")
```

Comment:
This was an ok advice someone gave, but because you had a `.json` file, `Data.Yaml.TH.decodeFile` would have been better, since you could have just made:
```haskell
magicMap :: MagicMap
magicMap = $$(decodeFile "data/magic.json")
```
Which would have also done the `parseJSON/eitherDecode'` part at compile time.

But someone also rightfully said that this package would do better with as few dependencies as possible. And aeson being very chunky, just turning the map into a Haskell file felt like the most optimal way of representing the data in this case.